lib/rubocop/cop/rspec/no_let.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+# TODO: when finished, run `rake generate_cops_documentation` to update the docs
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for usage of `let` blocks in specs.
+      #
+      # This cop can be configured with the option `AllowSubject` which
+      # will configure the cop to only register offenses on explicit calls to
+      # `let` and not calls to `subject`
+      #
+      # @example
+      #   # bad
+      #   describe MyClass do
+      #     let(:foo) { [] }
+      #     it { expect(foo).to be_empty }
+      #   end
+      #
+      #   describe MyClass do
+      #     subject(:foo) { [] }
+      #     it { expect(foo).to be_empty }
+      #   end
+      #
+      #   # good
+      #   describe MyClass do
+      #     it do
+      #       foo = []
+      #       expect(foo).to be_empty
+      #     end
+      #   end
+      #
+      # @example with AllowSubject configuration
+      #
+      #   # rubocop.yml
+      #   # RSpec/NoLet:
+      #   #   AllowSubject: true
+      #
+      #   # bad
+      #   describe MyClass do
+      #     let(:foo) { [] }
+      #     it { expect(foo).to be_empty }
+      #   end
+      #
+      #   # good
+      #   describe MyClass do
+      #     subject(:foo) { [] }
+      #     it { expect(foo).to be_empty }
+      #   end
+      #
+      #   describe MyClass do
+      #     it do
+      #       foo = []
+      #       expect(foo).to be_empty
+      #     end
+      #   end
+      #
+      class NoLet < Cop
+        MSG = 'Avoid using `let` and `subject` ' \
+              '– use a method call or local variable instead.'
+
+        def_node_matcher :let?, <<~PATTERN
+          (send nil? :let ...)
+        PATTERN
+
+        def_node_matcher :subject?, <<~PATTERN
+          (send nil? :subject ...)
+        PATTERN
+
+        def on_send(node)
+          return unless let?(node) || subject?(node)
+          return if subject?(node) && allow_subject?
```

Comment:
If you swap those two lines, the `|| subject?` becomes redundant.

Reply:
@pirj I'm not sure I understand this one. I tried this and it didn't flag `subject` calls:

```rb
return if subject?(node) && allow_subject?
return unless let?(node)
```

Reply:
Sorry, my bad. The way you've written it is better than `return unless let?(node) || (subject?(node) && allow_subject?)`.